pi.py

```diff
@@ -1 +1,578 @@
+    #   and Python PLCL code at http://romps.berkeley.edu/papers/pubdata/2016/lcl/lcl.py
+    #
+    CHI=TP/(1669.0-122.0*RH-TP)
+    PLCL=PP*(RH**CHI)
+    
+    # Initial default values before loop
+    CAPED=0
+    TOB=T[0]
+    IFLAG=1
+    # Values to help loop
+    NCMAX=0
+    jmin=int(1e6)
+    
+    #
+    #   ***  Begin updraft loop   ***
+    #
+
+    # loop over each level in the profile
+    for j in range(nlvl):
+        
+        # jmin is the index of the lowest pressure level evaluated in the loop
+        jmin=int(min([jmin,j]))
+    
+        #
+        #   *** Calculate Parcel quantities BELOW lifted condensation level   ***
+        #
+        if (P[j] >= PLCL):
+            # Parcel temperature at this pressure
+            TG=TP*(P[j]/PP)**(RD/CPD)
+            # Parcel Mixing ratio
+            RG=RP
+            # Parcel and Environmental Density Temperatures at this pressure (E94, EQN. 4.3.1 and 6.3.7)
+            TLVR=TG*(1.+RG/EPS)/(1.+RG)
+            TVENV=T[j]*(1.+R[j]/EPS)/(1+R[j])
+            # Bouyancy of the parcel in the environment (Proxy of E94, EQN. 6.1.5)
+            TVRDIF[j,]=TLVR-TVENV
+            
+        #
+        #   *** Calculate Parcel quantities ABOVE lifted condensation level   ***
+        # 
+        else:
+            
+            # Initial default values before loop
+            TGNEW=T[j]
+            TJC=T[j]-273.15
+            ES=6.112*np.exp(17.67*TJC/(243.5+TJC))
+            RG=EPS*ES/(P[j]-ES)
+            
+            #
+            #   ***  Iteratively calculate lifted parcel temperature and mixing   ***
+            #   ***                ratio for reversible ascent                    ***
+            #
+            
+            # set loop counter and initial condition
+            NC=0
+            TG=0
+
+            # loop until loop converges or bails out
+            while ((np.abs(TGNEW-TG)) > 0.001):
+            
+                # Parcel temperature and mixing ratio during this iteration
+                TG=TGNEW
+                TC=TG-273.15
+                ENEW=6.112*np.exp(17.67*TC/(243.5+TC))
+                RG=EPS*ENEW/(P[j]-ENEW)
+                
+                # increase iteration count in the loop
+                NC=NC+1
+                
+                #
+                #   ***  Calculate estimates of the rates of change of the entropy    ***
+                #   ***           with temperature at constant pressure               ***
+                #
+
+                ALV=ALV0+CPVMCL*TC
+                SL=(CPD+RP*CL+ALV*ALV*RG/(RV*TG*TG))/TG
+                EM=RG*P[j]/(EPS+RG)
+                # (last term vanishes with saturation, i.e. RH=1)
+                SG=(CPD+RP*CL)*np.log(TG)-RD*np.log(P[j]-EM)+ALV*RG/TG
+                # convergence speed (AP, step in entropy fraction) varies as a function of 
+                # number of iterations
+                if (NC < 3):
```

Comment:
Could you not change the gradient multiplication constant based on the magnitude of the gradient?

Reply:
You could! This might speed things up. Please feel free to give it a try. I haven't needed anything like the increased speed you have, but a branch to work on this could be warranted.